Geometry_Engine/Create/PlanarSurface.cs

```diff
@@ -55,31 +57,98 @@ public static PlanarSurface PlanarSurface(ICurve externalBoundary, List<ICurve>
-                    return null;
-                }
+            //---------------Coplanar-Internal-Boundaries-------------------//
+            count = internalBoundaries.Count;
+
+            Plane p = externalBoundary.IFitPlane();
+            internalBoundaries = internalBoundaries.Where(x => x.IIsInPlane(p)).ToList();
+
+            if (internalBoundaries.Count != count)
+            {
+                Reflection.Compute.RecordWarning("At least one of the internal boundaries is not coplanar with the external edge curve and has been ignored on creation of the planar surface.");
+            }
+
+            //--------------Unsupported-Internal-Boundaries-Warning---------//
+            if (internalBoundaries.Any(x => x is NurbsCurve || x is Ellipse))
+                Reflection.Compute.RecordWarning("At least one of the internal boundaries is a NurbsCurve or Ellipse and has not been checked for validity on creation of the planar surface.");
+
+            //--------------Self-Intersecting-Internal-Boundaries-----------//
+            count = internalBoundaries.Count;
+
+            internalBoundaries = internalBoundaries.Where(x => x is NurbsCurve || !x.IIsSelfIntersecting()).ToList();
+
+            if (internalBoundaries.Count != count)
+            {
+                Reflection.Compute.RecordWarning("At least one of the internal boundaries is self-intersecting and has been ignored on creation of the planar surface.");
+            }
+
+            //--------------Overlapping-Internal-Boundaries-----------------//
+            count = internalBoundaries.Count;
+
+            internalBoundaries = internalBoundaries.Where(x => x is NurbsCurve || x is Ellipse)
+                                .Concat(internalBoundaries.Where(x => !(x is NurbsCurve) && !(x is Ellipse)).BooleanUnion()).ToList();
+
+            if (internalBoundaries.Count != count)
+            {
+                Reflection.Compute.RecordWarning("At least one of the internalBoundaries was overlapping another one. BooleanUnion was used to resolve it.");
+            }
+
+            //--------------Unsupported-External-Boundary-------------------//
+            if (externalBoundary is NurbsCurve || externalBoundary is Ellipse)
+            {
+                Reflection.Compute.RecordWarning("External boundary is a nurbs curve or Ellipse. Necessary checks to ensure validity of a planar surface based on nurbs curve cannot be run, therefore correctness of the surface boundaries is not guaranteed.");
+                // External done
+                return new PlanarSurface(externalBoundary, internalBoundaries);
+            }
+
+            //--------------SelfIntersecting-External-Boundary--------------//
+            if (externalBoundary.IIsSelfIntersecting())
+            {
+                Reflection.Compute.RecordError("The provided external boundary is self-intersecting.");
+                return null;
```

Comment:
Hmm, I would put that right after the first two external boundary checks - then we have all 4 external checks at the beginning.

Reply:
Or, after seeing previous comment, this refers only to self-intersection, which def could be moved to the beginning.